pyiron_contrib/atomistics/atomicrex/structure_list.py

```diff
@@ -26,33 +26,49 @@ class FlattenedStructureContainer:
     Structures are stored together with an identifier that should be unique.
     The class has to be initialized with the number of structures and the total number of atoms in all structures.
     This allows to preallocate the necessary arrays and therefore prevents reallocations.
-
-    TODO: Alternatively preallocate arrays with large default values (f.e. 1e6 atoms) then use resize to shrink them?
-    """    
+    """
     def __init__(self, num_structures=0, num_atoms=0):
-        self.num_structures = num_structures
-        self.num_atoms = num_atoms
+        self._num_structures_alloc = num_structures
+        self._num_atoms_alloc = num_atoms
```

Comment:
I use num_structures and num_atoms in some other places in the atomicrex code, f.e. to iterate over the structures, so I wouldn't change their name.

Reply:
Maybe it is a good idea to have _num_structures_alloc for internal reallocations and num_structures for other uses like indexing and resizing before storing to hdf. When reading from hdf we allocate using the actual numbers and when structures are added they can be used as internal helper variables

Reply:
Fair enough, what about I rename `current_structure_index -> num_structures` and `current_atom_index -> num_atoms`?

Reply:
current_structure_index and current_atom_index are used to store the position in the structure and positions array after adding a structure, while num_structures and num_atoms store the total number of structures resp. atoms in the container and I think both are necessary for the functionality.

Reply:
But immediately after adding a structure the number of structures is the same as the index for the next structure or am I missing something?  Same for the atoms.

I see the scenarios:
1. immediately after creation of the container `num_structures=0`, `_num_structures_alloc=100`.  There's no structures inside yet and the first structures should be written at index 0, `current_structure_index=0`
2. after some structures have been added `num_structures=n`, `_num_structures_alloc=m>n`. The first `n` values of say, `self.cells` are taken, so the next one should go to index `n`, so `current_structure_index=n`.

Reply:
I can't think of a scenario where there wouldn't be an easy workaround that needs only 2 of those 3,
but considering clarity of names and _ naming convention I still think we should have all 3.
Then num_structures allows to iterate over all the structures, current_structure can be used to index into the array
and _num_structures_alloc is used for internal resizing and the cost to store 2 more ints is marginal cosidering that we store arrays with possibly millions of entries.
This also allows for some possibly nice features, f.e. we could set current_structure_index to the corresponding index when using get_structure(), which then allows to use it in arrays that store additional data.
For example in the atomicrex class I could index into my fit_property arrays using it then.

Alternatively I would leave out the _num_structures_alloc and use the num_structures for internal resizing, but then we need to  be super careful to always set it back to the correct number after reallocations, which probably gets much worse if this is used together with other flattened properties, like I do to store the fit_properties.

Reply:
Fair enough, if it makes other things easier, I'll add it back in.